meta-mender-core/classes/mender-ubimg.bbclass

```diff
@@ -0,0 +1,107 @@
+inherit mender-install
+
+# Class that creates an UBI image with an Mender layout
+
+# The UBI volyume scheme is:
+#    ubi0: first rootfs, active
+#    ubi1: second rootfs, inactive, mirror of first,
+#           available as failsafe for when some update fails
+#    ubi2: persistent data partition
+
+
+########## CONFIGURATION START - you can override these default
+##########                       values in your local.conf
+
+IMAGE_TYPEDEP_ubimg_append = "ubifs"
+
+########## CONFIGURATION END ##########
+
+inherit image
+inherit image_types
+
+IMAGE_DEPENDS_ubimg += " mtd-utils-native"
+
+IMAGE_CMD_ubimg () {
+    # For some reason, logging is not working correctly inside IMAGE_CMD bodies,
+    # so wrap all logging in these functions that also have an echo. This won't
+    # prevent warnings from being hidden deep in log files, but there is nothing
+    # we can do about that.
+    ubimg_warning() {
+        echo "$@"
+        bbwarn "$@"
+    }
+    ubimg_fatal() {
+        echo "$@"
+        bbfatal "$@"
+    }
+
+    mkdir -p "${WORKDIR}"
+
+    # Workaround for the fact that the image builder requires this directory,
+    # despite not using it. If "rm_work" is enabled, this directory won't always
+    # exist.
+    mkdir -p "${IMAGE_ROOTFS}"
+
+    REMAINING_SIZE=$(expr ${MENDER_STORAGE_TOTAL_SIZE_MB} - \
+         ${MENDER_DATA_PART_SIZE_MB} - 50)
+
+    CALC_ROOTFS_SIZE=$(expr $REMAINING_SIZE / 2)
```

Comment:
This has recently been moved to be a global calculation in `mender-install.bbclass`. It probably needs some adjustments there.